### PR TITLE
Unreserve braced enum variants in value namespace

### PR DESCRIPTION
**Pull request by petrochenkov**
With this PR braced enum variants (`enum E { V { /*...*/ } }`) no longer take a slot in value namespace, so the special case mentioned in the note in https://github.com/rust-lang/rfcs/blob/master/text/1506-adt-kinds.md#braced-structs is removed.

Report - https://github.com/rust-lang/rust/pull/103578#issuecomment-1292594900.

### CODE REVIEW

**Comment by petrochenkov** [comment]
### Change description for the lang team.

Tuple structs/variants and unit structs/variants occupy slots in two namespaces, like described in [RFC 1506 "Clarify the relationships between various kinds of structs and variants"](https://github.com/rust-lang/rfcs/blob/master/text/1506-adt-kinds.md).
```rust
struct S; // type + value (constructor constant)
struct S(u8); // type + value (constructor function)
struct S { field: u8 } // only type

enum E {
    V, // type + value (constructor constant)
    V(u8), // type + value (constructor function)
    V { field: u8 }, // type + value (???)
}
```

In addition to that braced variants (but not structs) occupy an additional slot in value namespace.
The value it refers to is unusable, it produces an error when referred to from an expression or pattern (it can be reexported around though, otherwise `use` wouldn't be usable with braced variants at all).

This special case for enum variants was introduced in early 2016 in https://github.com/rust-lang/rust/pull/30882 as future proofing.
During the last 6 years not specific plans for putting these names into use were proposed.
Also we now have editions to introduce these names in the future if such proposals appear.

So this PR removes these dummy names from value namespace thus making the language more consistent and simplifying the compiler.

File: compiler/rustc_middle/src/hir/map/mod.rs [hunk 1]
```diff
@@ -245,15 +245,12 @@ impl<'hir> Map<'hir> {
...
-                // FIXME(eddyb) is this even possible, if we have a `Node::Ctor`?
-                assert_ne!(variant_data.ctor_hir_id(), None);
-
                 let ctor_of = match self.find(self.get_parent_node(hir_id)) {
                     Some(Node::Item(..)) => def::CtorOf::Struct,
                     Some(Node::Variant(..)) => def::CtorOf::Variant,
                     _ => unreachable!(),
                 };
-                DefKind::Ctor(ctor_of, def::CtorKind::from_hir(variant_data))
+                DefKind::Ctor(ctor_of, variant_data.ctor_kind().unwrap())
```
Comment (nagisa) [inline compiler/rustc_middle/src/hir/map/mod.rs:253]: Consider `unwrap_or_else(|| bug!(...))`?

**Review by nagisa — COMMENTED** [review]
This generally looks okay, sans the diagnostics regressions, but those would be out of scope for me to vet and somebody working on diagnostics would need to review those for how fine is it to let them through in whatever state they are at the time this becomes ready for landing.

One thing worth doing now is a perf run though.

**Comment by petrochenkov** [comment]
>the diagnostics regressions

I made some changes to keep the diagnostics more or less the same.
Keeping all the suggestions/labels require porting some logic from main resolution pass (rustc_resolve) to type-relative resolution (rustc_hir_typeck), which is a more invasive change not directly related to unreserving braced enum variants.

**Comment by petrochenkov** [comment]
>One thing worth doing now is a perf run though.

@bors try @rust-timer queue

**Comment by nikomatsakis** [comment]
@rfcbot reviewed

**Comment by nagisa** [comment]
r=me, needs a rebase.

**Pull request merged**